### PR TITLE
ResNet50 model - end-to-end

### PR DESCRIPTION
**Pull request by nhasabni**
Opening up PR for review/feedback while I work on the benchmark.

### CODE REVIEW

**Review by rengolin — COMMENTED** [review]
Thanks! This looks good. Just a few comments...

First, I think we just need one example of each network (no batchnorm) in this repo, as this is about lowering MLIR, not necessarily all the complications of all models. We can test batch-norm in some other tests later. The more involved e2e examples should go in the e2e repositories.

Also, not all ops have CHECK lines. Some (convs) check for xsmm calls, others (max pool) for the unmodified linalg op (when we can't lower), that's great! But some (like the ReLUs) don't have any checks. It would be good to know if we missed any, so should add them too.

Finally, some blocks (bias add, padding and broadcast) don't have checks either, but those may fuse with surrounding ops. It would be nice to have either a CHECK line for the xsmm/linalg call, or a comment saying "This op fuses with the previous/next", so that we know we don't expect to see anything from that code.

**Comment by nhasabni** [comment]
Putting this PR on hold right now as we are planning to move this model to end-to-end repo.

**Comment by chelini** [comment]
> Putting this PR on hold right now as we are planning to move this model to end-to-end repo.

For me landing this is fine. We can experiment with pack/unpack propagation and tpp mapping.

**Comment by rengolin** [comment]
> For me landing this is fine. We can experiment with pack/unpack propagation and tpp mapping.

I don't mind merging just the "without batchnorm" either.

**Comment by nhasabni** [comment]
Al right. I will address the review comments.

**Comment by nhasabni** [comment]
> Thanks! This looks good. Just a few comments...
> 
> First, I think we just need one example of each network (no batchnorm) in this repo, as this is about lowering MLIR, not necessarily all the complications of all models. We can test batch-norm in some other tests later. The more involved e2e examples should go in the e2e repositories.
> 
> Also, not all ops have CHECK lines. Some (convs) check for xsmm calls, others (max pool) for the unmodified linalg op (when we can't lower), that's great! But some (like the ReLUs) don't have any checks. It would be good to know if we missed any, so should add them too.
> 
> Finally, some blocks (bias add, padding and broadcast) don't have checks either, but those may fuse with surrounding ops. It would be nice to have either a CHECK line for the xsmm/linalg call, or a comment saying "This op fuses with the previous/next", so that we know we don't expect to see anything from that code.

I don't see ReLUs getting lowered into XSMM calls.

For this
```
%27 = tensor.empty() : tensor<1x112x112x64xf32>
  %28 = linalg.generic {indexing_maps = [#map1, #map1, #map1], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%6, %cst_16 : tensor<1x112x112x64xf32>, tensor<1x112x112x64xf32>) outs(%27 : tensor<1x112x112x64xf32>) {
  ^bb0(%in: f32, %in_34: f32, %out: f32):
    %1591 = arith.maxf %in, %in_34 : f32
    linalg.yield %1591 : f32
  } -> tensor<1x112x112x64xf32>
```

I see

```
%alloc_4 = memref.alloc() {alignment = 64 : i64} : memref<1x112x112x64xf32>
scf.parallel (%arg2) = (%c0) to (%c112) step (%c1) {
  %subview_291 = memref.subview %alloc_3[0, %arg2, 0, 0] [1, 1, 112, 64] [1, 1, 1, 1] : memref<1x112x112x64xf32> to memref<1x1x112x64xf32, strided<[802816, 7168, 64, 1], offset: ?>>
  %subview_292 = memref.subview %63[0, %arg2, 0, 0] [1, 1, 112, 64] [1, 1, 1, 1] : memref<1x112x112x64xf32> to memref<1x1x112x64xf32, strided<[802816, 7168, 64, 1], offset: ?>>
  %subview_293 = memref.subview %alloc_4[0, %arg2, 0, 0] [1, 1, 112, 64] [1, 1, 1, 1] : memref<1x112x112x64xf32> to memref<1x1x112x64xf32, strided<[802816, 7168, 64, 1], offset: ?>>
  linalg.generic {indexing_maps = [#map1, #map1, #map1], iterator_types = ["parallel", "parallel"]} ins(%subview_291, %subview_292 : memref<1x1x112x64xf32, strided<[802816, 7168, 64, 1], offset: ?>>, memref<1x1x112x64xf32, strided<[802816, 7168, 64, 1], offset: ?>>) outs(%subview_293 : memref<1x1x112x64xf32, strided<[802816, 7168, 64, 1], offset: ?>>) {
      ^bb0(%in: f32, %in_294: f32, %out: f32):
        %296 = arith.maxf %in, %in_294 : f32
        linalg.yield %296 : f32
  }    
  scf.yield
}    
```

**Review by rengolin — APPROVED** [review]

**Comment by rengolin** [comment]
Nit: I'd just drop the "without-batchnorm" in the file names as we probably won't care about it ever in this project... Up to you.
